File: SQL/Select/SQLSelectFields.vb

' ___________________________________________________
'
'  © Hi-Integrity Systems 2010. All rights reserved.
'  www.hisystems.com.au - Toby Wicks
' ___________________________________________________
'

Option Strict On
Option Explicit On

Namespace SQL

    Public Class SQLSelectFields
        Implements Collections.Generic.IEnumerable(Of SQLSelectField)

        Private pobjFieldNames As New Collections.Generic.List(Of SQLSelectField)

        Public Sub New()

        End Sub

        Public Function Add() As SQLSelectField

            Return Add(String.Empty, String.Empty, 0, Nothing)

        End Function

        Public Function Add( _
            ByVal strFieldName As String) As SQLSelectField

            Return Add(strFieldName, String.Empty, 0, Nothing)

        End Function

        Public Function Add( _
            ByVal strFieldName As String, _
            ByVal eAggregateFunction As SQL.AggregateFunction) As SQLSelectField

            Return Add(strFieldName, String.Empty, eAggregateFunction, Nothing)

        End Function

        Public Function Add( _
            ByVal strFieldName As String, _
            ByVal strAlias As String, _
            ByVal eAggregateFunction As SQL.AggregateFunction) As SQLSelectField

            Return Add(strFieldName, strAlias, eAggregateFunction, Nothing)

        End Function

        Public Function Add( _
            ByVal strFieldName As String, _
            ByVal strAlias As String, _
            ByVal objTable As SQLSelectTable) As SQLSelectField

            Return Add(strFieldName, strAlias, 0, objTable)

        End Function

        Public Function Add( _
            ByVal strFieldName As String, _
            ByVal objTable As SQLSelectTable) As SQLSelectField

            Return Add(strFieldName, String.Empty, 0, objTable)

        End Function

        Public Function Add( _
            ByVal strFieldName As String, _
            ByVal strAlias As String, _
            ByVal eAggregateFunction As SQL.AggregateFunction, _
            ByVal objTable As SQLSelectTable) As SQLSelectField

            If eAggregateFunction = AggregateFunction.None Then
                Return Add(New SQLFieldExpression(objTable, strFieldName), strAlias)
            Else
                Return Add(New SQLFieldAggregateExpression(objTable, strFieldName, eAggregateFunction), strAlias)
            End If

        End Function

        Public Function Add(ByVal objExpression As SQLExpression) As SQLSelectField

            Return Add(objExpression, String.Empty)

        End Function

        Public Function Add(ByVal objExpression As SQLExpression, ByVal strAlias As String) As SQLSelectField

            Dim objSQLField As New SQLSelectField(objExpression)
            objSQLField.Alias = strAlias
            pobjFieldNames.Add(objSQLField)

            Return objSQLField

        End Function

        Public Sub Add(ByVal objFieldNames() As String)

            For intIndex As Integer = 0 To objFieldNames.Length - 1
                Me.Add(objFieldNames(intIndex))
            Next

        End Sub

        Default Public ReadOnly Property Item(ByVal strFieldName As String) As SQLSelectField
            Get

                Return Me(FieldNameIndex(strFieldName))

            End Get
        End Property

        Default Public ReadOnly Property Item(ByVal intIndex As Integer) As SQLSelectField
            Get

                Return pobjFieldNames.Item(intIndex)

            End Get
        End Property

        Public ReadOnly Property Count() As Integer
            Get

                Return pobjFieldNames.Count()

            End Get
        End Property

        Public Sub Clear()

            pobjFieldNames.Clear()

        End Sub

        Friend ReadOnly Property SQL(ByVal eConnectionType As Database.ConnectionType) As String
            Get

                Dim strSQL As String = String.Empty

                If Me.Count = 0 Then
                    strSQL = "*"
                Else
                    For intIndex As Integer = 0 To Me.Count - 1
                        strSQL &= Me.Item(intIndex).SQL(eConnectionType)
                        If intIndex <> Me.Count - 1 Then
                            strSQL &= ", "
                        End If
                    Next
                End If

                Return strSQL

            End Get
        End Property

        Public Function Exists(ByVal strFieldName As String) As Boolean

            Return FieldNameIndex(strFieldName) >= 0

        End Function

        Public Sub Delete(ByRef objSelectField As SQLSelectField)

            If Not pobjFieldNames.Contains(objSelectField) Then
                Throw New IndexOutOfRangeException
            End If

            pobjFieldNames.Remove(objSelectField)
            objSelectField = Nothing

        End Sub

        Private Function FieldNameIndex(ByVal strFieldName As String) As Integer

            strFieldName = strFieldName.Trim()

            For intIndex As Integer = 0 To Me.Count - 1
                If TypeOf Me.Item(intIndex).Expression Is SQLFieldExpression Then
                    Dim objFieldExpression As SQLFieldExpression = DirectCast(Me.Item(intIndex).Expression, SQLFieldExpression)
                    If String.Compare(strFieldName, objFieldExpression.Name, ignoreCase:=True) = 0 Then
                        Return intIndex
                    End If
                End If
            Next

            Return -1

        End Function

        Private Function GetEnumerator() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator

            Return pobjFieldNames.GetEnumerator

        End Function

        Private Function GetEnumerator1() As System.Collections.Generic.IEnumerator(Of SQLSelectField) Implements System.Collections.Generic.IEnumerable(Of SQLSelectField).GetEnumerator

            Return pobjFieldNames.GetEnumerator

        End Function

    End Class

End Namespace

File: SQL/Select/SQLSelectFields.cs
// _________________________________________________________________________
//
//  © Hi-Integrity Systems 2010. All rights reserved.
//  www.hisystems.com.au - Toby Wicks
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//	    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// _________________________________________________________________________
//

using System.Collections;
using System;
using System.Data;
using System.Collections.Generic;

namespace DatabaseObjects.SQL
{
	public class SQLSelectFields : IEnumerable<SQLSelectField>
	{
		private List<SQLSelectField> pobjFieldNames = new List<SQLSelectField>();
			
		public SQLSelectFields()
		{
		}
			
		public SQLSelectField Add()
		{
			return Add(string.Empty, string.Empty, 0, null);
		}
			
		public SQLSelectField Add(string strFieldName)
		{
			return Add(strFieldName, string.Empty, 0, null);
		}
			
		public SQLSelectField Add(string strFieldName, SQL.AggregateFunction eAggregateFunction)
		{
			return Add(strFieldName, string.Empty, eAggregateFunction, null);
		}
			
		public SQLSelectField Add(string strFieldName, string strAlias, SQL.AggregateFunction eAggregateFunction)
		{
			return Add(strFieldName, strAlias, eAggregateFunction, null);
		}
			
		public SQLSelectField Add(string strFieldName, string strAlias, SQLSelectTable objTable)
		{
			return Add(strFieldName, strAlias, 0, objTable);
		}
			
		public SQLSelectField Add(string strFieldName, SQLSelectTable objTable)
		{
			return Add(strFieldName, string.Empty, 0, objTable);
		}
			
		public SQLSelectField Add(string strFieldName, string strAlias, SQL.AggregateFunction eAggregateFunction, SQLSelectTable objTable)
		{
			if (eAggregateFunction == AggregateFunction.None)
				return Add(new SQLFieldExpression(objTable, strFieldName), strAlias);
			else
				return Add(new SQLFieldAggregateExpression(objTable, strFieldName, eAggregateFunction), strAlias);
		}
			
		public SQLSelectField Add(SQLExpression objExpression)
		{
			return Add(objExpression, string.Empty);
		}
			
		public SQLSelectField Add(SQLExpression objExpression, string strAlias)
		{
			SQLSelectField objSQLField = new SQLSelectField(objExpression);
			objSQLField.Alias = strAlias;
			pobjFieldNames.Add(objSQLField);
				
			return objSQLField;
		}
			
		public void Add(string[] objFieldNames)
		{
			for (int intIndex = 0; intIndex < objFieldNames.Length; intIndex++)
				this.Add(objFieldNames[intIndex]);
		}
			
		public SQLSelectField this[string strFieldName]
		{
			get
			{
				return this[FieldNameIndex(strFieldName)];
			}
		}
			
		public SQLSelectField this[int intIndex]
		{
			get
			{
				return pobjFieldNames[intIndex];
			}
		}
			
		public int Count
		{
			get
			{
				return pobjFieldNames.Count;
			}
		}
			
		public void Clear()
		{
			pobjFieldNames.Clear();
		}
			
		public bool Exists(string strFieldName)
		{
			return FieldNameIndex(strFieldName) >= 0;
		}
			
		public void Delete(ref SQLSelectField objSelectField)
		{
			if (!pobjFieldNames.Contains(objSelectField))
				throw new IndexOutOfRangeException();
				
			pobjFieldNames.Remove(objSelectField);
			objSelectField = null;
		}
			
		private int FieldNameIndex(string strFieldName)
		{
			for (int intIndex = 0; intIndex < this.Count; intIndex++)
			{
				if (this[intIndex].Expression is SQLFieldExpression)
				{
					SQLFieldExpression objFieldExpression = (SQLFieldExpression) (this[intIndex].Expression);
					if (string.Compare(strFieldName, objFieldExpression.Name, true) == 0)
						return intIndex;
				}
			}
				
			return -1;
		}
			
		public System.Collections.IEnumerator GetEnumerator()
		{
			return pobjFieldNames.GetEnumerator();
		}
			
		IEnumerator<SQLSelectField> IEnumerable<SQLSelectField>.GetEnumerator()
	    {
		    return pobjFieldNames.GetEnumerator();
	    }

	    System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
	    {
			return pobjFieldNames.GetEnumerator();
	    }
	}
}